Sources/XiEditor/MarkerBar.swift

```diff
@@ -32,17 +32,17 @@ class Marker {
 class MarkerBar: NSScroller {
     var markers: [Marker] = []
     var backgroundColor: NSColor = NSColor.red
-    var parent: EditViewController!
-    var markerLayer = CAShapeLayer()
-    var overlayScrollerLayer: CALayer?
+    let markerLayer = CAShapeLayer()
+    var overlayScrollerLayer: CALayer!
+    weak var markerDelegate: EditViewController?
 
     override func viewWillDraw() {
         super.viewWillDraw()
 
         // last child layer will be used for drawing the overlay scroller
-        overlayScrollerLayer = self.layer?.sublayers?.last
-        overlayScrollerLayer?.addObserver(self, forKeyPath: "opacity", options: .new, context: nil)
-        self.layer?.addSublayer(markerLayer)
+        overlayScrollerLayer = self.layer!.sublayers!.last
+        overlayScrollerLayer.addObserver(self, forKeyPath: "opacity", options: .new, context: nil)
```

Comment:
do we need to set this observer each time? Does the layer change? In any case it's important when adding an observer that you also remove it before the object gets deallocated; see [the docs](https://developer.apple.com/documentation/objectivec/nsobject/1412787-addobserver#) for more detail.

it also isn't obvious that we need to hold on to a reference to this layer. Maybe I'm missing something?

Reply:
We need the layer later again to determine the width of the scrollbar (it is wider when mouse cursor is on top) when drawing the markers. It would be possible to use `self.layer!.sublayers!.last` though but I thought `overlayScrollerLayer` would be more descriptive.